### ISSUE TITLE
MP3s make music, not videos

### ISSUE THREAD

**Issue by FlameSoulis** [Yellow-Dog-Man/Resonite-Issues#216]
### Describe the bug?

Uploading an MP3 will offer to upload as a video file rather than an audio file.

### To Reproduce

1. Upload your favorite MP3 file.
2. Be greeted by the upload dialog.
3. Don't see Audio as an option.
4. Be disapointed.

### Expected behavior

Uploading an MP3 offers to upload as an audio file.

### Screenshots

https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/6835457/c552e4a1-3344-4b88-baec-329288427e0a

![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/6835457/9b86051c-1f8e-4854-a6d6-47347510aac4)

### Resonite Version Number

Beta 2023.10.13.743

### What Platforms does this occur on?

Windows, Linux

### What headset if any do you use?

Desktop, Index, Quest

### Additional Context

The MP3 file format is no longer requiring a license to use, and has been a very popular question on why music files aren't working in Resonite, when those files were MP3. Resolving this issue could potentially resolve 'false bugs' on 'audio not working/importing.'

### Reporters

flamesoulis (U-Flame-Soulis)

**Comment by maksim789456** [comment]
The same story with `m4a, aac, ac3, aif, aiff, opus` audio formats

FrooxEnigne recognize this audio formats as video, it's shows at logs on early FrooxEngine init
`0:06:31.836 ( -1 FPS)	Supported audio formats: wav, wave, flac, fla, ogg, aiff, aif, aifc`
`0:06:31.836 ( -1 FPS)	Supported video formats: mp4, mpeg, avi, mov, mpg, mkv, flv, webm, mts, 3gp, bik, m2v, m2s, wmv, m3u8, m3u, pls, m4a, mp3, mpeg3, aac, ac3, aif, aiff, ape, au, it, mka, mod, mp1, mp2, opus, s3m, sid, w64, wma, xm, nsf, nsfe, gbs, vgm, vgz, spc, gym`

**Comment by gameboycjp** [comment]
For context on why this happens, the library that handles audio doesn't support these formats, so the engine defers to video player libraries instead. That's paraphrasing  on what I've been told, at least. Giving the user the option to convert to a supported audio format would be nice as a stopgap, but full support would be awesome too.

**Comment by Frooxius** [comment]
This one is a bit tricky. The goal is to treat them as audio clip assets, but right now we lack a good miltiplatform library to achieve this, that wouldn't add much development/dependency burden.

We tried a few C# implementations, but they had severe issues (screeching audio with some files, locking up with others).

We can look into native libs, where the support is better, but it also adds more development burden, as we have to cross compile on multiple platforms.

We'd probably look into one's that support a large number of formats, so it's not just specifically for MP3.

We're open to suggestions for libs that we can, look at. Ideally pure C#. We already tried NLayer and MP3Sharp.

**Comment by FlameSoulis** [comment]
Wouldn't a humble update to assimp resolve the issue?

**Comment by ProbablePrime** [comment]
> Wouldn't a humble update to assimp resolve the issue?

AssImp doesn't support MP3. (https://assimp-docs.readthedocs.io/en/latest/about/introduction.html)

**Comment by FlameSoulis** [comment]
Oh... hmm... I thought that was handling the media stuff as well. My bad.

**Comment by gentlecolts** [comment]
is ffmpeg not a viable solution to this?

**Comment by ProbablePrime** [comment]
FFmpeg might work as an auto-convert, but we'd have to integrate that carefully, thanks for the suggestion.

**Comment by Frooxius** [comment]
FFmpeg wouldn't help here either, that would act as a "media player" too, just like libVLC does.

**Comment by sveken** [comment]
In regards to OPUS. Doesn't reso already support opus for audio streams with an opus library?  It would be the best format to replace OGG vorbis as it is the successor lossy audio format to it.

**Comment by Frooxius** [comment]
It's definitely not a bug, but this issue shouldn't be "blocked". The only thing blocking it is us deciding to invest time into it.

There are some libraries for this and part of the work would be re-searching and integrating those, which we can do at any time - it just makes it a bigger chunk of work.

If someone from the community wants to do this research and find good options, that would make this smaller chunk of work on our end and more likely to prioritize it.

**Comment by KyuubiYoru** [comment]
Would it be possible to add an option to convert unsupported audio formats like MP3 to supported formats (OGG, FLAC, WAV) in the import dialog using FFmpeg?

**Comment by FlameSoulis** [comment]
> Would it be possible to add an option to convert unsupported audio formats like MP3 to supported formats (OGG, FLAC, WAV) in the import dialog using FFmpeg?

Hmm... so, VLC can do conversations... is this extended also to libVLC?

**Comment by Frooxius** [comment]
> Would it be possible to add an option to convert unsupported audio formats like MP3 to supported formats (OGG, FLAC, WAV) in the import dialog using FFmpeg?

I wouldn't introduce FFmpeg as dependency just for this - it's pretty big. 

However in general, it's possible - we could even use the C# library, since it's main issue seems to be that it freaks out when seeking, so it could be a possible approach, but would lead to slower import.

> Hmm... so, VLC can do conversations... is this extended also to libVLC?
Unfortunately not with the version we're using (UMP).

**Comment by Nytra** [comment]
Apparently CSCore can do MP3 and that library is already used in Resonite

**Comment by Nytra** [comment]
There are a couple of cross-platform options:

- FFmpeg wrapper like https://github.com/tomaszzmuda/Xabe.FFmpeg
- BASS Audio (Requires paid license for commerical use) https://www.un4seen.com/

**Comment by gentlecolts** [comment]
https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/216#issuecomment-1791943679
see above, ffmpeg has been mentioned and I have doubts that the team would incorporate a non-open system

**Comment by Frooxius** [comment]
@Nytra We've already talked about FFmpeg extensively above. I *strongly* recommend reading the rest of this thread first.

We're aware of BASS Audio too, but like you said, it's commercial, so that's an investment.

**Comment by AwesomeTornado** [comment]
I found a good mp3 library, it is written in C, and designed to be small and fast.
https://github.com/lieff/minimp3
It would need a wrapper, but it looks very promising.

It is licensed under CC0, making it incredibly easy to modify and incorporate in any capacity, and seems to have all the necessary features to convert to some other usable format on import, or to simply play the audio directly.

**Comment by Frooxius** [comment]
@AwesomeTornado Thanks! I've taken a quick look. It's potentially promising - it's simple enough API that it should be pretty easy to make a wrapper and it shouldn't be hard to compile either.

One main concern I have right now with that library are the seeking capabilities - it seems that it's based on the byte positions in the buffer, which might not correspond to actual time. We need the seeking to be able to seek to a precise time/sample-position and I'm not sure if it's trivially achievable with this library.

**Comment by AwesomeTornado** [comment]
You are correct, the library does not have the ability to seek to a timestamp.

There are two solutions that I see.
1: Convert MP3 files to some other format with full support on import
2: Implement a custom timestamp search feature.

For number one, this should be easy to implement, and would prevent MP3 files from causing problems, but it would require extra processing time on import. In my opinion, this is the best option

For number two, each file would have to be indexed on import or on the fly in order to support seeking. File duration is not included in the metadata, and would have to be calculated by summing the playtime of each mp3 frame. Additionally, each frame is not guaranteed to be the same bitlength (or sample speed I think), which means that seeking to a new time is **not** as simple as finding a relative position. As an example, halfway through the MP3 audio is not always the same as halfway through the MP3 file. 

I have a lot of ideas about how to optimize MP3 file indexing/searching for memory and speed, but I am not sure that they are entirely relevant at this point.

**Comment by gentlecolts** [comment]
I will also add that it's fairly normal for game engines to have their own internal formats for media, doing conversion would be a pretty typical thing as long as the original is recoverable

**Comment by Frooxius** [comment]
Neither of those solutions is workable for this.

1. While this is technically possible, this adds complexity, explicit steps and won't work all cases. It slows down the import process and won't work with assets that you don't explicitly bring in (e.g. from https, streams and so on).
2. Too much effort for this, I'm not willing to invest that.

Implementing our own system for this should be the last resort when we exhaust possible libraries. Both Mp3Sharp and NLayer support this. Many native libraries do too.

**Comment by Zozokasu** [comment]
I am not familiar with the audio API, but I found an audio library that seems to meet the requirements.
https://github.com/mackron/miniaudio
- Low-level API with access to raw audio data
- mp3 support
- Cross-platform
- public domain or MIT No Attribution

**Comment by Frooxius** [comment]
Thank you! That one looks pretty good. Are there any existing C# wrappers?

**Comment by Stellanora64** [comment]
@Frooxius I found this, but I'm not sure what's the state it's in https://www.nuget.org/packages/MiniaudioSharp

**Comment by Frooxius** [comment]
Hmm it looks like it's no longer actively developed: https://github.com/Starpelly/MiniaudioSharp

**Comment by Zozokasu** [comment]
I found this one, but it says .NET 8.0 and C# 12 in the dependencies, and very work in progress.
https://github.com/MrBrixican/Foster.Audio
